e2etest/bgp_test.go

```diff
@@ -80,8 +71,94 @@ var _ = ginkgo.Describe("BGP", func() {
 		} else {
 			ginkgo.By("checking connectivity to its external VIP with docker")
 		}
-		err = wgetRetry(useDocker, address)
+		err := wgetRetry(useDocker, address)
 		framework.ExpectNoError(err)
 	})
 
+	ginkgo.Context("metrics", func() {
+		var controllerPod *corev1.Pod
+		var speakerPods []*corev1.Pod
+
+		ginkgo.BeforeEach(func() {
+			pods, err := cs.CoreV1().Pods("metallb-system").List(context.Background(), metav1.ListOptions{
+				LabelSelector: "component=controller",
+			})
+			framework.ExpectNoError(err)
+			framework.ExpectEqual(len(pods.Items), 1, "More than one controller found")
+			controllerPod = &pods.Items[0]
+
+			speakers, err := cs.CoreV1().Pods("metallb-system").List(context.Background(), metav1.ListOptions{
+				LabelSelector: "component=speaker",
+			})
+			framework.ExpectNoError(err)
+			speakerPods = make([]*corev1.Pod, 0)
+			for _, item := range speakers.Items {
+				speakerPods = append(speakerPods, &item)
+			}
+		})
+
+		ginkgo.It("should be exposed by the controller", func() {
+			peerAddr := "172.18.0.5:179" // TODO replace this when we create the config locally
+			poolName := "dev-env-bgp"    // TODO replace this when we create the config locally
+			ginkgo.By("checking the metrics when no service is added")
+			controllerMetrics, err := metrics.ForPod(controllerPod, controllerPod)
+			framework.ExpectNoError(err)
+			validateGaugeValue(0, "metallb_allocator_addresses_in_use_total", map[string]string{"pool": poolName}, controllerMetrics)
+			validateGaugeValue(256, "metallb_allocator_addresses_total", map[string]string{"pool": poolName}, controllerMetrics)
+
+			for _, speaker := range speakerPods {
+				ginkgo.By(fmt.Sprintf("checking speaker %s", speaker.Name))
+
+				speakerMetrics, err := metrics.ForPod(controllerPod, speaker)
+				framework.ExpectNoError(err)
+				validateGaugeValue(1, "metallb_bgp_session_up", map[string]string{"peer": peerAddr}, speakerMetrics)
+				validateGaugeValue(0, "metallb_bgp_announced_prefixes_total", map[string]string{"peer": peerAddr}, speakerMetrics)
+			}
+
+			ginkgo.By("creating a service")
+			svc := createServiceWithBackend(cs, f.Namespace.Name) // Is a sleep required here?
```

Comment:
Check this todo? I don't think so as `svc, err := jig.CreateLoadBalancerService(timeout, tweakServicePort())` will timeout and fail if not created. You can confirm by looking at the funciton in jig https://github.com/kubernetes/kubernetes/blob/b0bc8adbc2178e15872f9ef040355c51c45d04bb/test/e2e/framework/service/jig.go#L245 and from what I see it waits with `return j.WaitForLoadBalancer(timeout)`

Reply:
Yes, what I don't know is how long it will take for the metrics to appear.
After the service (and the endpoint) is created, it may take some time to metallb to start producing the metrics and we may scrape them too early (causing the test to flake).
I'd say let's keep an eye on this knowing that this could be the cause.